### PR TITLE
Add helper for building element sides that minimizes construction

### PR DESCRIPTION
**Pull request by loganharbour**
The capability to pass in a possibly exiting element in the form of a `unique_ptr` to `Elem::build_side_ptr()` was added in the past. If the element that is passed in is of the correct type required to build a side, no new allocation is made and the necessary information is transferred in place. 

This helper, the `ElemSideBuilder`, stores a cache that is indexed by `ElemType` and "builds" element sides given an element and a side in a way that requires the construction of an element for a given type only once. With this also comes `Elem::side_type()`, which returns the `ElemType` of a given side and is needed for indexing into the cache in `ElemSideBuilder.`

Additionally, I used this new helper in a few places and also cleaned up some side building where appropriate.

### CODE REVIEW

File: include/geom/elem_side_builder.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include "libmesh/elem.h"
+
+namespace libMesh
+{
+
+/**
+ * Helper for building element sides that minimizes the construction
+ * of new elements.
+ *
+ * TODO: add more info
```
Comment (roystgnr) [inline include/geom/elem_side_builder.h:32]: Indeed.

File: include/geom/elem_side_builder.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+{
+public:
+  /**
+   * Builds an element side for side \p s of element \p elem.
+   */
+  Elem & operator(const Elem & elem, const unsigned int s);
+
+private:
+  /// Proxy element cache for building side pointers; indexed by ElemType
+  std::vector<std::unique_ptr<Elem>> _cached_elems;
```
Comment (roystgnr) [inline include/geom/elem_side_builder.h:44]: This is going to end up quite sparse, isn't it?  I can't think of a better way to do it, though; typically for containers this small the relative overhead of anything but a vector/array is huge.

File: tests/geom/elem_test.C [hunk 3]
```diff
@@ -139,14 +140,37 @@ public:
...
+
+  void test_elem_side_builder()
+  {
+    ElemSideBuilder cache;
+    for (const auto & elem : _mesh->active_local_element_ptr_range())
+      for (const auto s : elem->side_index_range())
+      {
+        const auto side = elem->build_side_ptr(s);
+        const auto & cached_side = cache(*elem, s);
+        CPPUNIT_ASSERT_EQUAL(e1.type(), e2.type());
```
Comment (roystgnr) [inline tests/geom/elem_test.C:159]: I'm guessing this'll be fixed by the time I'm done typing this comment...

File: include/geom/elem_side_builder.h (hunk 2 above)
Comment (loganharbour) [reply include/geom/elem_side_builder.h:44]: Yep. Only other option here is if we want to go static with an array; same sparsity issue though

**Comment by loganharbour** [comment]
Barring any failures on my part, this should be good to go. Take a look at the doctring for `ElemSideBuilder` and let me know what you think.

**Comment by roystgnr** [comment]
Docs look good to me.  Let's wait for @jwpeterson to check it out before merging, though.  I've got a few other things (both the outstanding PRs and one other branch still in progress) that I'm hoping to get into the next Moose libMesh submodule update, myself, so I don't want to rush that too too much.

**Comment by loganharbour** [comment]
Are there tests for inf elements anywhere?

**Comment by loganharbour** [comment]
Few more things, of which objections to any are okay with me:

- Applied `ElemSideBuilder` to the examples that build sides; this may be overkill and it may be more appropriate to just use a single `std::unique_ptr<const Elem>`, but I figured it also wasn't an idea to expose users to the class in the examples
- Applied `ElemSideBuilder` to a few methods in libmesh that _could_ be building element sides for multiple types
- Used a single element in a few places for building sides where it was appropriate

**Comment by jwpeterson** [comment]
> Let's wait for @jwpeterson to check it out before merging, 

After @roystgnr explained the issue to me in terms of Prism elements, where you are switching back and forth between different geometric side types, the idea makes sense to me. So, assuming that the tests start passing, I'd be OK with merging? Would you mind adding a description to the PR as well?

File: src/mesh/mesh_tools.C [hunk 4]
```diff
@@ -366,9 +366,7 @@ void MeshTools::find_boundary_nodes (const MeshBase & mesh,
     for (auto s : elem->side_index_range())
       if (elem->neighbor_ptr(s) == nullptr) // on the boundary
         {
-          std::unique_ptr<const Elem> side = elem->build_side_ptr(s);
```
Comment (roystgnr) [inline src/mesh/mesh_tools.C:369]: What?  How did we never get an unused variable warning for this?  Is the non-trivial destructor enough for the compiler to have to give up and say "maybe they're doing some fancy RAII thing"?

Comment (loganharbour) [reply src/mesh/mesh_tools.C:369]: Right?! I had the same thought!

**Comment by loganharbour** [comment]
> So, assuming that the tests start passing, I'd be OK with merging? Would you mind adding a description to the PR as well?

Yep - assuming this is passing, this is good to go. It does require a once-over from one of you to confirm that the places that I added `ElemSideBuilder` are appropriate.

In addition to @roystgnr - I did make a change to `ElemSideBuilder` that allows for building const and non-const sides that should be looked at (uses a little const casting)

File: examples/miscellaneous/miscellaneous_ex9/augment_sparsity_on_interface.C [hunk 5]
```diff
@@ -35,23 +36,26 @@ void AugmentSparsityOnInterface::mesh_reinit ()
...
 
+  // To avoid extraneous allocation when getting element side vertex averages
+  std::unique_ptr<const Elem> side_elem;
+
   for (const auto & elem : _mesh.active_element_ptr_range())
     for (auto side : elem->side_index_range())
       if (elem->neighbor_ptr(side) == nullptr)
         {
           if (_mesh.get_boundary_info().has_boundary_id(elem, side, _crack_boundary_lower))
-            {
```
Comment (loganharbour) [inline examples/miscellaneous/miscellaneous_ex9/augment_sparsity_on_interface.C:43]: Gah... stupid spaces

**Comment by loganharbour** [comment]
This isn't completely ready. I'm going to add infinite elem tests to `elem_test.C`

**Comment by loganharbour** [comment]
Good to go.

File: src/geom/cell_inf_hex8.C [hunk 6]
```diff
@@ -67,7 +67,7 @@ const unsigned int InfHex8::edge_nodes_map[InfHex8::num_edges][InfHex8::nodes_pe
 const unsigned int InfHex8::edge_sides_map[InfHex8::num_edges][2] =
   {
     {0, 1}, // Edge 0
-    {1, 2}, // Edge 1
+    {0, 2}, // Edge 1
```
Comment (roystgnr) [inline src/geom/cell_inf_hex8.C:70]: Come again?  I could imagine having screwed up this table somehow (or having copied a mistake in the diagram in cell_inf_hex8.h?), but not like this.  An InfHex vertex has three edges connected to it.  There's no way node 0 is part of 4 edges.

Comment (loganharbour) [reply src/geom/cell_inf_hex8.C:70]: Sound logic. This is actually a map that I implemented :-) 

I'll look into it

**Comment by roystgnr** [comment]
That `edge_sides_map` weirdness was the only problem I could find.

File: src/geom/cell_inf_hex8.C (hunk 6 above)
Comment (loganharbour) [reply src/geom/cell_inf_hex8.C:70]: Oh - this is edge -> side map. Has nothing to do with nodes. It's valid.

Comment (loganharbour) [reply src/geom/cell_inf_hex8.C:70]: Actually, hold up. Need to correct it in the other classes, too.

**Comment by loganharbour** [comment]
Done, again. Consolidated the edge_sides_map for the cell classes, moving the map to the base classes instead. I duplicated the same map for all of the specialized cells when it was unnecessary.

**Review by roystgnr — APPROVED** [review]

**Comment by roystgnr** [comment]
@jwpeterson - looks like 2434108 is necessary + sufficient to backport onto the release branch

**Pull request merged**